File: scala/pal/src/main/scala/pal/ScaffoldWindowWorkers.scala
package pal

import scala.collection.immutable.VectorMap
import Expr.{FALSE, TRUE}
import Ref.PREVIOUS
import ScaffoldCircuit.{choose, conjunction as AND, disjunction as OR, neg}

/** GS matcher and proper-prefix flags on the original input-node stream.
  *
  * Every actual input character advances the global raw end once. Frozen views
  * are cursor snapshots into that stream: reverse views read just left of their
  * boundary cursor. They never copy, pad, or append synthetic input symbols.
  */
trait WindowWorkerLike {
  def output: Expr
  def mode: Value[String]
  def flagPool: Option[FlagPackets]
  def flags: Option[FlagStack]
  def arrive(): Unit
  def resetFlags(enabled: Expr): Unit
  def start(enabled: Expr): Unit
  def mark(enabled: Expr): Unit
  def service(): Unit
  def commit(): Unit
}

final class WindowWorker(val circuit: Circuit, val prefix: String, val quantum: Int,
                         val program: Program, names: Seq[String], val isFlags: Boolean = false)
  extends WindowWorkerLike {
  if (quantum < 1) {
    throw new IllegalArgumentException("a positive finite service quantum is required")
  }

  val readers: ReaderLiveness = GsHeadLiveness.analyzeReaders(program)
  val readerNames: Vector[String] = Vector.tabulate(readers.registers)(index => s"$prefix.data$index")
  private val movement = program.code.iterator.flatMap { instruction =>
    instruction.event match {
      case Event.Move(moves) => moves.iterator.map(move => math.abs(move.delta))
      case _ => Iterator.empty
    }
  }.maxOption.getOrElse(1)
  val hKey: String = prefix + ".h"
  val distances: WindowLiveDistances = new WindowLiveDistances(
    circuit, program, names, 1 + 2 * quantum * movement, prefix + ".distance",
    if (isFlags) { Vector(hKey) } else { Vector.empty }, availabilityDistance = false)
  val table: ROM = ScaffoldRom.controllerTable(
    program, names, Some(readers), readerNames, batched = true, augment = Some(distances.augmentRows))
  val stream: WindowStream = new WindowStream(
    circuit, Vector(prefix + ".begin", prefix + ".end") ++ readerNames,
    3 + quantum * movement, prefix + ".stream")
  val begin: WindowHead = stream.heads(prefix + ".begin")
  val end: WindowHead = stream.heads(prefix + ".end")
  val data: VectorMap[String, WindowHead] = VectorMap.from(readerNames.map(name => name -> stream.heads(name)))
  var reverse: VectorMap[String, Expr] = VectorMap.from(readerNames.map { name =>
    name -> circuit.get(PREVIOUS, name + ".reverse", Vector(false, true), false).eqTo(true)
  })
  val pcKey: String = prefix + ".pc"
  val modeKey: String = prefix + ".mode"
  var pc: Value[Int] = circuit.get(PREVIOUS, pcKey, program.code.indices.toVector, program.start)
  var mode: Value[String] = circuit.get(PREVIOUS, modeKey, Vector("idle", "run", "done"), "idle")
  var output: Expr = FALSE
  val flagPool: Option[FlagPackets] = if (isFlags) { Some(new FlagPackets(circuit, quantum, prefix + ".packets")) } else { None }
  val flags: Option[FlagStack] = flagPool.map(pool => new FlagStack(pool, prefix + ".output"))

  def arrive(): Unit = {
    end.move(1)
    distances.loadOne(TRUE)
    if (isFlags) { distances.values.add(hKey, 1) }
  }

  private def rawStart(head: String, source: WindowHead, reversed: Boolean, enabled: Expr): Unit = {
    val name = readerNames(readers.colors(head))
    data(name).copyFrom(source, enabled)
    reverse = reverse.updated(name, choose(enabled, if (reversed) { TRUE } else { FALSE }, reverse(name)))
  }

  def resetFlags(enabled: Expr): Unit = {
    val result = flags.getOrElse(throw new IllegalArgumentException("only flag workers have a segment builder"))
    begin.copyFrom(end, enabled)
    distances.values.reset(distances.lengthKey, enabled)
    distances.values.reset(hKey, enabled)
    result.clear(enabled)
    mode = Value.select(enabled, Value.constant("idle"), mode)
  }

  def start(enabled: Expr): Unit = {
    if (isFlags) {
      circuit.require(neg(mode.eqTo("run")), enabled)
      rawStart("Origin", begin, reversed = false, enabled)
      rawStart("TextOrigin", end, reversed = true, enabled)
      rawStart("OriginalEnd", begin, reversed = true, enabled)
      distances.initializeValues(VectorMap(
        ("Lower", "Upper") -> Some((hKey, -1)),
        ("Origin", "OriginalEnd") -> Some((distances.lengthKey, -1)),
        ("Origin", "Upper") -> Some((distances.lengthKey, -1)),
        ("OriginalEnd", "Lower") -> Some((hKey, 1)),
        ("OriginalEnd", "TextOrigin") -> Some((distances.lengthKey, 1)),
        ("OriginalEnd", "Upper") -> None), enabled)
      flags.foreach(_.clear(enabled))
    } else {
      rawStart("Origin", end, reversed = true, enabled)
      rawStart("Tail", end, reversed = false, enabled)
      distances.initializeValues(VectorMap(("Origin", "Tail") -> Some((distances.lengthKey, -1))), enabled)
    }
    pc = Value.select(enabled, Value.constant(program.start), pc)
    mode = Value.select(enabled, Value.constant("run"), mode)
  }

  def mark(enabled: Expr): Unit = distances.values.reset(hKey, enabled)

  private def orientation(which: Value[Any]): Expr = {
    OR(reverse.iterator.map { case (name, value) => AND(which.eqTo(name), value) }.toSeq*)
  }

  private def read(which: Value[Any], enabled: Expr): Value[Any] = {
    val head = stream.selectHead(which)
    head.offset = Bits.select(orientation(which), head.offset.add(-1), head.offset)
    head.read(enabled)
  }

  def service(): Unit = {
    val writer = flags.map(new PacketWriter(_))
    for (_ <- 0 until quantum) {
      step(mode.eqTo("run"), writer)
      Expr.clearExpressionCache()
    }
  }

  private def step(active: Expr, writer: Option[PacketWriter]): Unit = {
    val fields = table.read(pc.bits)
    val opcode = fields("op")
    val (equal, less) = distances.compare(fields)
    val symbolTest = AND(active, opcode.eqTo("symbols"))
    val a = read(fields("data_left"), symbolTest)
    val b = read(fields("data_right"), symbolTest)
    val available = stream.selectHead(fields("data_left")).available()
    val decision = choose(opcode.eqTo("equal"), equal,
      choose(opcode.eqTo("less"), less,
        choose(opcode.eqTo("available"), available, a.equal(b))))
    circuit.require(equal, AND(active, opcode.eqTo("assert_equal")))
    if (!isFlags) {
      val matched = AND(active, opcode.eqTo("match"))
      val bName = readerNames(readers.colors("B"))
      circuit.require(neg(data(bName).available()), matched)
      output = OR(output, matched)
    }
    distances.execute(fields, active)
    val source = stream.selectHead(fields("data_source"))
    val reverseSource = orientation(fields("data_source"))
    for ((name, head) <- data) {
      val rawDelta = fields("data_delta." + name)
      val delta = Value.select(reverse(name), rawDelta.map {
        case amount: Int => -amount
        case amount => throw new IllegalArgumentException(s"non-integer head delta: $amount")
      }, rawDelta)
      head.moveSelected(delta, active)
      val copying = AND(active, opcode.eqTo("copy"), fields("data_target").eqTo(name))
      head.copyFrom(source, copying)
      reverse = reverse.updated(name, choose(copying, reverseSource, reverse(name)))
    }
    if (isFlags) {
      writer.get.push(fields("bit").eqTo(true), AND(active, opcode.eqTo("flag")))
      mode = Value.select(AND(active, opcode.eqTo("halt")), Value.constant("done"), mode)
    } else {
      circuit.require(FALSE, AND(active, opcode.eqTo("halt")))
    }
    pc = Value.select(active, Value.select(decision, fields("yes").asInstanceOf[Value[Int]], fields("no").asInstanceOf[Value[Int]]), pc)
  }

  def commit(): Unit = {
    distances.commit()
    stream.commit()
    circuit.put(pcKey, pc)
    circuit.put(modeKey, mode)
    for ((name, bit) <- reverse) {
      circuit.put(name + ".reverse", Value.select(bit, Value.constant(true), Value.constant(false)))
    }
    flags.foreach(_.commit())
    circuit.put(prefix + ".matched", Value.select(output, Value.constant(true), Value.constant(false)), Vector(false, true), false)
  }
}

object ScaffoldWindowWorkers {
  def matcher(circuit: Circuit, prefix: String = "matcher", quantum: Int = 512): WindowWorker = {
    new WindowWorker(circuit, prefix, quantum, GsMatchHeads.compileMatcher(unit = false), GsMatchHeads.MATCH_HEADS)
  }
  def flags(circuit: Circuit, prefix: String = "flags", quantum: Int = 1024): WindowWorker = {
    new WindowWorker(circuit, prefix, quantum, GsDualFlags.compileDualFlags(unit = false), GsDualFlags.DUAL_HEADS, isFlags = true)
  }
}

File: docs/palindromes-in-peg/scaffold_window_workers.py
"""GS matcher and proper-prefix flags on the original input-node stream.

Every actual input character advances the global raw end once. Frozen views
are cursor snapshots into that stream: reverse views read just left of their
boundary cursor. They do not copy, pad, or append synthetic input symbols.
"""
from copy import copy

from gs_match_heads import compile_matcher, MATCH_HEADS
from gs_dual_flags import compile_dual_flags, DUAL_HEADS
from gs_head_liveness import analyze_readers
from scaffold_circuit import (Value, PREVIOUS, TRUE, FALSE, conjunction as AND,
                              disjunction as OR, neg, choose)
from scaffold_rom import controller_table
from scaffold_window_counter import Bits
from scaffold_window_live import WindowLiveDistances
from scaffold_window_stream import WindowStream
from scaffold_flag_packets import FlagPackets, FlagStack, PacketWriter
from symbolic_sca2peg import clear_expression_cache


class WindowWorker:
  def __init__(self, circuit, prefix, quantum, program, names, *, flags=False):
    if type(quantum) is not int or quantum < 1:
      raise ValueError("a positive finite service quantum is required")
    self.circuit, self.prefix, self.quantum, self.program = circuit, prefix, quantum, program
    self.is_flags = flags
    self.readers = analyze_readers(program)
    self.reader_names = tuple(f"{prefix}.data{i}" for i in range(self.readers.registers))
    movement = max((abs(delta) for event, _ in program.code if event[0] == "move"
                    for _, delta in event[1]), default=1)
    self.h_key = prefix + ".h"
    self.distances = WindowLiveDistances(circuit, program, names,
      1 + 2 * quantum * movement, prefix + ".distance", extra=(self.h_key,) if flags else (),
      availability_distance=False)
    self.table = controller_table(program, names, self.readers, self.reader_names,
                                  batched=True, augment=self.distances.augment_rows)
    self.stream = WindowStream(circuit, (prefix + ".begin", prefix + ".end", *self.reader_names),
                               3 + quantum * movement, prefix + ".stream")
    self.begin, self.end = (self.stream.heads[prefix + name] for name in (".begin", ".end"))
    self.data = {name: self.stream.heads[name] for name in self.reader_names}
    self.reverse = {name: circuit.get(PREVIOUS, name + ".reverse", (False, True), False).eq(True)
                    for name in self.reader_names}
    self.pc_key, self.mode_key = prefix + ".pc", prefix + ".mode"
    self.pc = circuit.get(PREVIOUS, self.pc_key, tuple(range(len(program.code))), program.start)
    self.mode = circuit.get(PREVIOUS, self.mode_key, ("idle", "run", "done"), "idle")
    self.output = FALSE
    if flags:
      self.flag_pool = FlagPackets(circuit, quantum, prefix + ".packets")
      self.flags = FlagStack(self.flag_pool, prefix + ".output")
    self.writer = None

  def arrive(self):
    self.end.move(1)
    self.distances.load_one(TRUE)
    if self.is_flags:
      self.distances.values.add(self.h_key, 1)

  def raw_start(self, head, source, reverse, enabled):
    name = self.reader_names[self.readers.colors[head]]
    self.data[name].copy_from(source, enabled)
    self.reverse[name] = choose(enabled, TRUE if reverse else FALSE, self.reverse[name])

  def reset_flags(self, enabled):
    if not self.is_flags: raise ValueError("only flag workers have a segment builder")
    self.begin.copy_from(self.end, enabled)
    self.distances.values.reset(self.distances.length_key, enabled)
    self.distances.values.reset(self.h_key, enabled)
    self.flags.clear(enabled)
    self.mode = Value.select(enabled, Value.constant("idle"), self.mode)

  def start(self, enabled):
    c, d = self.circuit, self.distances
    if self.is_flags:
      c.require(neg(self.mode.eq("run")), enabled)
      self.raw_start("Origin", self.begin, False, enabled)
      self.raw_start("TextOrigin", self.end, True, enabled)
      self.raw_start("OriginalEnd", self.begin, True, enabled)
      d.initialize_values({("Lower", "Upper"): (self.h_key, -1),
        ("Origin", "OriginalEnd"): (d.length_key, -1),
        ("Origin", "Upper"): (d.length_key, -1),
        ("OriginalEnd", "Lower"): (self.h_key, 1),
        ("OriginalEnd", "TextOrigin"): (d.length_key, 1),
        ("OriginalEnd", "Upper"): None}, enabled)
      self.flags.clear(enabled)
    else:
      self.raw_start("Origin", self.end, True, enabled)
      self.raw_start("Tail", self.end, False, enabled)
      d.initialize_values({("Origin", "Tail"): (d.length_key, -1)}, enabled)
    self.pc = Value.select(enabled, Value.constant(self.program.start), self.pc)
    self.mode = Value.select(enabled, Value.constant("run"), self.mode)

  def mark(self, enabled):
    self.distances.values.reset(self.h_key, enabled)

  def orientation(self, which):
    return OR(*(AND(which.eq(name), value) for name, value in self.reverse.items()))

  def read(self, which, enabled):
    head = self.stream.select_head(which)
    # The movement radius reserves a further unit for reverse reads.
    head.offset = Bits.select(self.orientation(which), head.offset.add(-1), head.offset)
    return head.read(enabled)

  def service(self):
    if self.is_flags:
      self.writer = PacketWriter(self.flags)
    for _ in range(self.quantum):
      self.step(self.mode.eq("run"))
      clear_expression_cache()

  def step(self, active):
    c, fields = self.circuit, self.table.read(self.pc.bits)
    opcode = fields["op"]
    equal, less = self.distances.compare(fields)
    symbol_test = AND(active, opcode.eq("symbols"))
    a, b = (self.read(fields[key], symbol_test) for key in ("data_left", "data_right"))
    available = self.stream.select_head(fields["data_left"]).available()
    decision = choose(opcode.eq("equal"), equal, choose(opcode.eq("less"), less,
               choose(opcode.eq("available"), available, a.equal(b))))
    c.require(equal, AND(active, opcode.eq("assert_equal")))
    if not self.is_flags:
      matched = AND(active, opcode.eq("match"))
      c.require(neg(self.data[self.reader_names[self.readers.colors["B"]]].available()), matched)
      self.output = OR(self.output, matched)
    self.distances.execute(fields, active)
    source = self.stream.select_head(fields["data_source"])
    reverse_source = self.orientation(fields["data_source"])
    for name, head in self.data.items():
      delta = fields["data_delta." + name]
      delta = Value.select(self.reverse[name], delta.map(lambda amount: -amount), delta)
      head.move_selected(delta, active)
      copying = AND(active, opcode.eq("copy"), fields["data_target"].eq(name))
      head.copy_from(source, copying)
      self.reverse[name] = choose(copying, reverse_source, self.reverse[name])
    if self.is_flags:
      self.writer.push(fields["bit"].eq(True), AND(active, opcode.eq("flag")))
      self.mode = Value.select(AND(active, opcode.eq("halt")), Value.constant("done"), self.mode)
    else:
      c.require(FALSE, AND(active, opcode.eq("halt")))
    self.pc = Value.select(active, Value.select(decision, fields["yes"], fields["no"]), self.pc)

  def finalize(self):
    self.distances.finalize()
    self.stream.finalize()
    self.circuit.put(self.pc_key, self.pc)
    self.circuit.put(self.mode_key, self.mode)
    for name, bit in self.reverse.items():
      self.circuit.put(name + ".reverse", Value.select(bit, Value.constant(True), Value.constant(False)))
    if self.is_flags:
      self.flags.finalize()
    self.circuit.put(self.prefix + ".matched", Value.select(self.output, Value.constant(True), Value.constant(False)),
                     (False, True), False)


def matcher(circuit, prefix="matcher", quantum=512):
  return WindowWorker(circuit, prefix, quantum, compile_matcher(unit=False), MATCH_HEADS)


def flags(circuit, prefix="flags", quantum=1024):
  return WindowWorker(circuit, prefix, quantum, compile_dual_flags(unit=False), DUAL_HEADS, flags=True)
